### PR TITLE
Implemented personal pricing for ecommerce

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Part of #1058 

#### What's this PR do?
Uses personal price when charging user.

#### How should this be manually tested?
Talk to me first to set up the Cybersource test environment variables

- Create a `TierProgram` for some program with some discount amount if you don't have one already. 
- Create a `FinancialAid` with status `approved` referencing that `TierProgram`.
-  Create a `CoursePrice` on some `CourseRun` which has `is_valid=True`. Make sure that the price is more than the discount you set. Set up `CourseRun` such that the enroll button shows up in the dashboard (soon to be replaced with a Pay button)
 - Click the enroll button. You should be directed to Cybersource. The total being requested should be equal to the price in `CoursePrice` minus the discount amount.
 - Go into the Django shell and verify that an `Order` is created for this course run, that it has `total_price_paid` equal to the discounted price, and that there is one `Line` in the `Order` whose `price` is also equal to the discounted price.

Then, test that negative or zero prices are rejected:
- Delete the `Order` that was created in the Django shell.
 - Set `CoursePrice.price` to a value lower or equal to the discount amount
 - Click the enroll button again. You should get a 500 error, and an `ImproperlyConfigured` exception in the Django console

#### What GIF best describes this PR or how it makes you feel?
7ktihtl

### CODE REVIEW

File: ecommerce/api.py [hunk 1]
```diff
@@ -88,7 +93,18 @@ def create_unfulfilled_order(course_id, user):
         Order: A newly created Order for the CourseRun with the given course_id
     """
     course_run = get_purchasable_course_run(course_id, user)
-    price = course_run.courseprice_set.get(is_valid=True).price
+
+    price_dict = get_course_price_for_learner(user, course_run.course.program)
+    price = price_dict['course_price']
```
Comment (giocalitri) [inline ecommerce/api.py]: Question: can you be sure of the existence of the `'course_price'` key?

File: ecommerce/api_test.py [hunk 2]
```diff
@@ -108,28 +129,61 @@ def test_already_purchased(self):
...
         assert ex.exception.args[0] == 'Course run {} is already purchased'.format(course_run.edx_course_key)
 
+    def test_less_or_equal_to_zero(self):
+        """
+        An order may not have a negative or zero price
+        """
+        course_run, user = create_purchasable_course_run()
+        price_obj = course_run.courseprice_set.get(is_valid=True)
+
+        for invalid_price in (0, -1.23):
```
Comment (giocalitri) [inline ecommerce/api_test.py]: Nit: being a tuple, can you please write it as `(0, -1.23,)`

**Comment by giocalitri** [comment]
Code looks good. Checking functionality

**Comment by giocalitri** [comment]
Not sure it is a bug related to this PR, but if I do not have a `FinancialAid` object at all for the use in the program, I can pay automatically the full price.

File: ecommerce/api.py (hunk 1 above)
Comment (noisecapella) [inline ecommerce/api.py]: Yes, the function has only one return statement and the key is defined in the dict there

**Comment by noisecapella** [comment]
We should fix that, I'll change the `get_course_price_for_learner` function to raise a `ValidationError` in that case. Our workflow should create a `FinancialAid` with a skipped status somehow before this API is called

**Comment by noisecapella** [comment]
Now that I think about it the check should be done in the checkout API, not in `get_course_price_for_learner`

**Comment by noisecapella** [comment]
Ok, the user should now not be able to purchase without at least one `FinancialAid` attached

**Comment by giocalitri** [comment]
I think that the same problem now exists if the FinancialAid object has a status that should not allow a purchase.
For example I do not expect that a FinancialAid with status `created` should me allow to make a payment.

**Comment by noisecapella** [comment]
That makes sense, I'll restrict it to the three final states

**Comment by noisecapella** [comment]
I also filtered on the user which I missed earlier. Should be ready for review again

File: ecommerce/api.py [hunk 3]
```diff
@@ -55,12 +64,24 @@ def get_purchasable_course_run(course_key, user):
...
 
+    if not FinancialAid.objects.filter(
+            tier_program__current=True,
+            tier_program__program__course__courserun=course_run,
+            user=user,
+            status__in=FinancialAidStatus.TERMINAL_STATUSES,
+    ).exists():
+        log.warning("Course run %s has no attached financial aid for user %s", course_key, get_social_username(user))
+        raise ValidationError(
+            "Course run {} does not have a current attached financial aid application".format(course_key)
```
Comment (giocalitri) [inline ecommerce/api.py]: nit: I could specify "valid" here, considering that there could be one in a different state, but feel free to leave as is

**Comment by giocalitri** [comment]
👍

**Comment by giocalitri** [comment]
I tested a discount equal to the maximum price and the the backend correctly raises `ImproperlyConfigured`.
This translates to a 500 error. Should we address this translating the `ImproperlyConfigured` to a proper JSON answer instead of an HTML?
screen shot 2016-09-30 at 11 54 52 am

**Comment by giocalitri** [comment]
In the meantime I confirm the 👍

**Pull request merged**